### PR TITLE
Add DCO

### PR DESCRIPTION
**Pull request by r0mainK**
Reacting to this: https://github.com/src-d/models/pull/4 I've added DCO to the commit message. You can pass it as arguments with `committer-name` and `committer-email` flags, or simply define them in `modelforgecfg.py` or ENV variables like for other "annoying" parameters.

### CODE REVIEW

**Comment by r0mainK** [comment]
@vmarkovtsev this is ready for review/merge. In the end I also added small changes in the `index.py` file, to manage the case where the pulled git repo does not have an index file. At first I wanted to just put it as a `warning`, since we could assume the repo does not contain the appropriate file because it's being initialized, but then thought it might be best for the user to initially commit and empty JSON file (containing just `{}`). If you think we should change I can simply add a boolean parameter to the `__init__` of GitIndex to catch the error if we're running the `init` command and fail otherwise

File: README.md [hunk 1]
```diff
@@ -40,6 +40,8 @@ BACKEND = "gcs"  # type of backend to use
 BACKEND_ARGS = "bucket='user_bucket.models',credentials='key.json'"  # all backend arguments 
 INDEX_REPO = "https://github.com/user/models"  # git repo for the index
 CACHE_DIR = "~/.cache/modelforge"  # default cache to use for the index
+COMMITTER_NAME = Random J Developer  # default name to use for the DCO
```
Comment (vmarkovtsev) [inline README.md]: We should take it from the global git config (`~/.gitconfig`). I bet dulwich is able to parse such files.

**Comment by vmarkovtsev** [comment]
@r0mainK So these two global vars are not needed - we should parse the git config file every time we commit.

Also it's better to extract the index changes into a separate PR.

**Comment by r0mainK** [comment]
@vmarkovtsev gonna check if `dulwich` implements that, fot the time being I extracted the other changes : https://github.com/src-d/modelforge/pull/37

File: modelforge/index.py [hunk 2]
```diff
@@ -177,7 +178,22 @@ def upload(self, cmd: str, meta: dict):
...
+                conf = ConfigFile.from_file(_in)
+                try:
+                    name = conf.get(b"user", b"name").decode()
+                    email = conf.get(b"user", b"email").decode()
+                    message += self.DCO_MESSAGE.format(name=name, email=email)
+                except KeyError:
+                    self._log.warning("Did not find name or email in %s, committing with no DCO.",
+                                      global_conf_path)
+        else:
+            self._log.warning("No global configuration file in %s, committing with no DCO.",
```
Comment (vmarkovtsev) [inline modelforge/index.py]: Better message: `"Global git configuration file %s does not exist, committing without DCO."`

File: modelforge/index.py [hunk 3]
```diff
@@ -177,7 +178,22 @@ def upload(self, cmd: str, meta: dict):
...
+        global_conf_path = os.path.expanduser("~/.gitconfig")
+        if os.path.exists(global_conf_path):
+            with open(global_conf_path, "br") as _in:
+                conf = ConfigFile.from_file(_in)
+                try:
+                    name = conf.get(b"user", b"name").decode()
+                    email = conf.get(b"user", b"email").decode()
+                    message += self.DCO_MESSAGE.format(name=name, email=email)
+                except KeyError:
+                    self._log.warning("Did not find name or email in %s, committing with no DCO.",
```
Comment (vmarkovtsev) [inline modelforge/index.py]: `with no` -> `without`

**Comment by vmarkovtsev** [comment]
To be merged cnce Travis is happy.

**Comment by smola** [comment]
Does this mean that users will be signing-off commits unknowingly? This may render sign-offs legally void unless the user is informed before the fact.

**Comment by r0mainK** [comment]
did not know that ^^" @smola would adding a disclaimer in the `README.md` and a `--no-dco` flag circumvent this ?

**Comment by smola** [comment]
I would say that we need some kind of explicit action by the user such as adding a flag, accepting an interactive prompt, setting a configuration file or something like that, but I'm not really sure.

@eiso Can we get some advice here?

**Comment by vmarkovtsev** [comment]
@smola good point, I didn't know either

**Comment by vmarkovtsev** [comment]
I guess the `--signoff` option must be added to make the DCO optional.

**Comment by vmarkovtsev** [comment]
Also configuration flag `ALWAYS_SIGNOFF` which is `False` by default.

**Comment by r0mainK** [comment]
okay making the change, will ping you once it is added

**Comment by smola** [comment]
@vmarkovtsev [IANAL](https://en.wikipedia.org/wiki/IANAL) but adding an explicit environment variable / CLI flag that is disabled by default looks on the safe side.

**Comment by r0mainK** [comment]
@vmarkovtsev added the feature

File: modelforge/configuration.py [hunk 4]
```diff
@@ -8,6 +8,7 @@
 BACKEND_ARGS = os.getenv("MODELFORGE_BACKEND_ARGS", "")
 INDEX_REPO = os.getenv("MODELFORGE_INDEX_REPO", "")
 CACHE_DIR = os.getenv("MODELFORGE_CACHE_DIR", os.path.expanduser("~/.cache"))
+ALWAYS_SIGNOFF = os.getenv("MODELFORGE_ALWAYS_SIGN_OFF", False)
```
Comment (vmarkovtsev) [inline modelforge/configuration.py]: Better to be consistent and have `MODELFORGE_ALWAYS_SIGNOFF`

File: modelforge/__main__.py [hunk 5]
```diff
@@ -32,6 +32,8 @@ def add_index_args(p):
                        help="Url of the remote Git repository.")
         p.add_argument("--cache", default=None,
                        help="Path to the folder where the Git repository will be cached.")
+        p.add_argument("-s", "--signoff", action="store_true",
+                       help="To add a DCO to the commit message.")
```
Comment (vmarkovtsev) [inline modelforge/__main__.py]: Copied from `git commit --help`:

>Add Signed-off-by line by the committer at the end of the commit log message. The meaning of a signoff depends on the project, but it typically certifies that committer has the rights to submit this work under the same license and agrees to a Developer Certificate of Origin (see http://developercertificate.org/ for more information).

File: modelforge/configuration.py (hunk 4 above)
Comment (r0mainK) [reply modelforge/configuration.py]: my mistake, copy pasted

File: modelforge/__main__.py (hunk 5 above)
Comment (r0mainK) [reply modelforge/__main__.py]: on it - going to solve merge conflicts as well

**Comment by r0mainK** [comment]
@vmarkovtsev solved the conflicts and did the changes you asked. I also modified the `README.md` and `doc/cmd.md`

**Pull request merged**